src/engine/util/webxr/raycaster.ts

```diff
@@ -0,0 +1,108 @@
+import {
+  Raycaster,
+  Vector3,
+  BufferGeometry,
+  LineBasicMaterial,
+  Line,
+  Matrix4,
+} from "three";
+
+/** ThreeJS Line, extended with an Update method */
+class SCLine extends Line {
+  constructor(
+    bufferGeo: THREE.BufferGeometry,
+    material: THREE.LineBasicMaterial
+  ) {
+    super(bufferGeo, material);
+  }
+  Update(): void {}
+}
+
+/**
+ * Sandcastle's own Raycaster; WebXR-friendly syntactic sugar over threeJS's Raycaster
+ * @extends Raycaster
+ */
+export class SCRaycaster extends Raycaster {
+  _originObject: THREE.Mesh | THREE.Group;
+  _target: THREE.Mesh | Array<THREE.Mesh> | THREE.Box3;
+  _direction: THREE.Vector3;
+  _isRecursive: boolean;
+  _near: number;
+  _far: number;
+  _visualizedRaycast: THREE.Line | undefined;
+  _isTargetArray: boolean;
+  _isTargetBox3: boolean;
+  _tempMatrix: THREE.Matrix4;
+  /**
+   * Construct a Sandcastle Raycaster.
+   * @param originObject - object to raycast from
+   * @param target - object(s) to raycast to. Can be a mesh, a mesh array or a Box3.
+   * @param [direction] - The normalized direction vector that gives direction to the ray.
+   * @param [isRecursive] -  If true, it also checks all descendants. Otherwise it only checks intersection with the object. Default is true.
+   * @param [near] - All results returned are further away than near. Near can't be negative. Default value is 0.1.
+   * @param [far] - All results returned are closer than far. Far can't be lower than near. Default value is 10.
+   */
+  constructor(
+    originObject: THREE.Mesh | THREE.Group,
+    target: THREE.Mesh | Array<THREE.Mesh> | THREE.Box3, // due to original polymorphism in THREE.Raycaster(), see below
+    direction: THREE.Vector3 = new Vector3(0, 0, -1),
+    isRecursive: boolean = true,
+    near: number = 0.1,
+    far: number = 10
+  ) {
+    super();
+    if (!originObject.parent || originObject.parent.type != "Scene") {
+      throw new Error("Error: the raycasting object is not in the scene!");
+    }
+
+    this._tempMatrix = new Matrix4();
+    this._originObject = originObject;
+    this._target = target;
+    this._direction = direction;
+    this._isRecursive = isRecursive;
+    this._near = near;
+    this._far = far;
+    this._visualizedRaycast = undefined;
+    this._isTargetArray = Array.isArray(this._target); // cache check because it will impact every frame
```

Comment:
You could move `this._isTargetArray` and `this._isTargetBox3` to local variables in getInterctions. That would reduce the private variable count and keep the issues around caching scoped to that function.

Reply:
true - unfortunately it would also mean running `Array.isArray()` and `hasOwnProperty` checks every frame (the typical frequency of getIntersections() which is something I tried to avoid by caching `target`'s type in the constructor.